mantidproject/mantid
# New algorithm: SofTwoThetaTOF

This PR introduces a new algorithm `SofTwoThetaTOF` which averages a direct geometry time-of-flight workspace over constant 2theta producing an S(2theta, TOF) workspace. The instrument scientists are intending to use this type of workspace for diagnostic purposes at the IN5 instrument at the ILL.

This PR also makes the generation of the `.par` file in `GenerateGroupingPowder` optional as the file is not needed for `SofTwoThetaTOF` and would complicate the cleanup process.

**Report to:** Jacques Ollivier/ILL.

**To test:**

Ensure that the unit test data directories are included in Mantid's data search directories. Run the script below.
```python
ws = DirectILLCollectData('ILL/IN5/104007.nxs')
stt = SofTwoThetaTOF(ws)
```

This should produce a workspace where the elastic channels are aligned between 7000-8000us and the vertical axis has units of Degrees (2theta).

Filling the input workspace with a constant number should reproduce the constant in the output workspace except where the detectors don't cover specific 2theta and TOF regions:
```python
import numpy as np
ws = DirectILLCollectData('ILL/IN5/104007.nxs')
for i in range(ws.getNumberHistograms()):
    ws.dataY(i).fill(2.)
    ws.dataE(i).fill(np.sqrt(2.))
stt = SofTwoThetaTOF(ws)
```

Fixes #23808.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
